34 add generic parsing logic

TNO-SlaFleur opened the pull request:
1. Added a parser which can find out the message type automatically and parse the message accordingly. Instead of doing all kinds of reflective black magic, I went for generating the mapping from message type to the class which should be used for parsing. By generating this mapping the result is fast and simple and regenerating should be breeze.

2. Allow to also parse from a dict: `ValidateValuesMixin.from_dict()`

3. Add in and fix lint and typecheck to ci and tox while also fixing any outstanding issues.

victorgarcia98 commented on src/s2python/s2_parser.py:
```diff
@@ -0,0 +1,91 @@
+    def parse_dict_as_any_message(self, message_dict: dict) -> S2Message:
+        message_type = self.get_message_type_from_dict(message_dict)
+
+        if message_type not in TYPE_TO_MESSAGE_CLASS:
```
Nice!

This also covers for the case of not having the `message_type` key.

victorgarcia98 commented on src/s2python/s2_parser.py:
```diff
@@ -0,0 +1,91 @@
+
+        :param message: The message as a JSON-formatted string.
+        :param as_message: The type of message that is expected within the `message`
+        :raises: S2ValidationError
```
This could also raise a `json.JSONDecodeError`.

TNO-SlaFleur replied:
Fixed!

victorgarcia98 commented on src/s2python/s2_parser.py:
```diff
@@ -0,0 +1,91 @@
+}
+
+
+class S2Parser:
```
Maybe, we could have a method that accepts both a str and a dict:

```python
def parse(self, message: str | dict) -> S2Message:
    if isintance(message, str):
        message = json.loads(message)
    
    message_type = self.get_message_type_from_dict(message)

    if message_type not in TYPE_TO_MESSAGE_CLASS:
        raise S2ValidationError(
            message,
            f"Unable to parse {message_type} as an S2 message. Type unknown.",
        )

    return TYPE_TO_MESSAGE_CLASS[message_type].parse_obj(message)    
```

TNO-SlaFleur replied:
Fixed!

victorgarcia98 reviewed (commented):
Nice feature! This is very useful to start creating objects from serialized data :smile:

TNO-SlaFleur commented:
@victorgarcia98 Apologies, I kinda went overboard. I added in lint and typecheck as well as I ran into some issues actually running the tools. I fixed all outstanding issues and turned on the tools in ci as well. All of these changes got stuck into this PR. Hope you don't mind, otherwise I can look into splitting them into 2. I did take care to change the original description of the PR to include these changes.

victorgarcia98 reviewed (approved):
Nice, this is very handy!

Also, thanks for fixing the typing issues! :smile:
